File: backend/app/routers/morph.py
"""Face morphing endpoints."""

from __future__ import annotations

import json
import time
from typing import List, Optional

from fastapi import APIRouter, File, Form, Request, UploadFile
from fastapi.responses import JSONResponse

from app.models.schemas import (
    ErrorCodes,
    ErrorDetail,
    ErrorResponse,
    ImageDimensions,
    MorphData,
    MorphResponse,
    ResponseMeta,
    StagedMorphData,
    StagedMorphResponse,
    StageImage,
)
from app.services.morphing import get_morphing_service
from app.utils.image import ImageValidationError, cv2_to_base64, validate_image

router = APIRouter(tags=["morph"])

DEFAULT_STAGES = [0.0, 0.25, 0.5, 0.75, 1.0]


@router.post("/morph", response_model=None)
async def morph_faces(
    request: Request,
    current_image: Optional[UploadFile] = File(None),
    ideal_image: Optional[UploadFile] = File(None),
    progress: float = Form(0.5),
):
    """
    Morph between two face images.

    Args:
        request: FastAPI request object
        current_image: Current face image (JPEG/PNG, max 10MB)
        ideal_image: Ideal/target face image (JPEG/PNG, max 10MB)
        progress: Morphing progress (0.0 = current, 1.0 = ideal)

    Returns:
        Morphed image as base64-encoded PNG
    """
    start_time = time.time()
    request_id = getattr(request.state, "request_id", "unknown")

    # Validate progress parameter
    if progress < 0.0 or progress > 1.0:
        return JSONResponse(
            status_code=400,
            content=ErrorResponse(
                error=ErrorDetail(
                    code=ErrorCodes.VALIDATION_ERROR,
                    message="Progress must be between 0.0 and 1.0",
                )
            ).model_dump(),
        )

    # Validate current_image is provided
    if current_image is None or current_image.filename == "":
        return JSONResponse(
            status_code=400,
            content=ErrorResponse(
                error=ErrorDetail(
                    code=ErrorCodes.VALIDATION_ERROR,
                    message="current_image is required",
                )
            ).model_dump(),
        )

    # Validate ideal_image is provided
    if ideal_image is None or ideal_image.filename == "":
        return JSONResponse(
            status_code=400,
            content=ErrorResponse(
                error=ErrorDetail(
                    code=ErrorCodes.VALIDATION_ERROR,
                    message="ideal_image is required",
                )
            ).model_dump(),
        )

    try:
        # Read and validate images
        current_data = await current_image.read()
        ideal_data = await ideal_image.read()

        if len(current_data) == 0:
            return JSONResponse(
                status_code=400,
                content=ErrorResponse(
                    error=ErrorDetail(
                        code=ErrorCodes.VALIDATION_ERROR,
                        message="current_image is required",
                    )
                ).model_dump(),
            )

        if len(ideal_data) == 0:
            return JSONResponse(
                status_code=400,
                content=ErrorResponse(
                    error=ErrorDetail(
                        code=ErrorCodes.VALIDATION_ERROR,
                        message="ideal_image is required",
                    )
                ).model_dump(),
            )

        # Validate and load images
        _, cv2_current = validate_image(current_data)
        _, cv2_ideal = validate_image(ideal_data)

        # Perform morphing
        morph_service = get_morphing_service()
        result = morph_service.morph(
            cv2_current,
            cv2_ideal,
            progress,
            img1_label="current",
            img2_label="ideal",
        )

        # Get dimensions
        h, w = result.shape[:2]

        # Convert to base64
        image_base64 = cv2_to_base64(result, "png")

        # Calculate processing time
        processing_time_ms = (time.time() - start_time) * 1000

        return MorphResponse(
            data=MorphData(
                image=image_base64,
                format="png",
                progress=progress,
                dimensions=ImageDimensions(width=w, height=h),
            ),
            meta=ResponseMeta(
                request_id=request_id,
                processing_time_ms=processing_time_ms,
            ),
        )

    except ImageValidationError as e:
        status_code_map = {
            ErrorCodes.IMAGE_TOO_LARGE: 413,
            ErrorCodes.INVALID_IMAGE_FORMAT: 400,
            ErrorCodes.FACE_NOT_DETECTED: 400,
            ErrorCodes.MULTIPLE_FACES: 400,
        }
        status_code = status_code_map.get(e.code, 400)

        return JSONResponse(
            status_code=status_code,
            content=ErrorResponse(
                error=ErrorDetail(
                    code=e.code,
                    message=e.message,
                    details=e.details,
                )
            ).model_dump(),
        )

    except Exception as e:
        print(f"Morph error: {e}")

        return JSONResponse(
            status_code=500,
            content=ErrorResponse(
                error=ErrorDetail(
                    code=ErrorCodes.PROCESSING_ERROR,
                    message="Failed to generate morphed image",
                )
            ).model_dump(),
        )


@router.post("/morph/stages", response_model=None)
async def morph_stages(
    request: Request,
    current_image: Optional[UploadFile] = File(None),
    ideal_image: Optional[UploadFile] = File(None),
    stages: Optional[str] = Form(None),
):
    """
    Generate multiple morphing stages between two face images.

    Args:
        request: FastAPI request object
        current_image: Current face image (JPEG/PNG, max 10MB)
        ideal_image: Ideal/target face image (JPEG/PNG, max 10MB)
        stages: JSON array of progress values (e.g., "[0, 0.25, 0.5, 0.75, 1.0]")

    Returns:
        Multiple morphed images at different progress levels
    """
    start_time = time.time()
    request_id = getattr(request.state, "request_id", "unknown")

    # Parse stages
    stage_values: List[float] = DEFAULT_STAGES
    if stages is not None and stages.strip():
        try:
            parsed = json.loads(stages)
            if not isinstance(parsed, list):
                return JSONResponse(
                    status_code=400,
                    content=ErrorResponse(
                        error=ErrorDetail(
                            code=ErrorCodes.VALIDATION_ERROR,
                            message="stages must be a JSON array of numbers",
                        )
                    ).model_dump(),
                )

            stage_values = [float(v) for v in parsed]

            # Validate all values are in range
            for v in stage_values:
                if v < 0.0 or v > 1.0:
                    return JSONResponse(
                        status_code=400,
                        content=ErrorResponse(
                            error=ErrorDetail(
                                code=ErrorCodes.VALIDATION_ERROR,
                                message="All stage values must be between 0.0 and 1.0",
                                details={"invalid_value": v},
                            )
                        ).model_dump(),
                    )

        except (json.JSONDecodeError, ValueError, TypeError):
            return JSONResponse(
                status_code=400,
                content=ErrorResponse(
                    error=ErrorDetail(
                        code=ErrorCodes.VALIDATION_ERROR,
                        message="stages must be a valid JSON array of numbers",
                    )
                ).model_dump(),
            )

    # Validate images
    if current_image is None or current_image.filename == "":
        return JSONResponse(
            status_code=400,
            content=ErrorResponse(
                error=ErrorDetail(
                    code=ErrorCodes.VALIDATION_ERROR,
                    message="current_image is required",
                )
            ).model_dump(),
        )

    if ideal_image is None or ideal_image.filename == "":
        return JSONResponse(
            status_code=400,
            content=ErrorResponse(
                error=ErrorDetail(
                    code=ErrorCodes.VALIDATION_ERROR,
                    message="ideal_image is required",
                )
            ).model_dump(),
        )

    try:
        # Read and validate images
        current_data = await current_image.read()
        ideal_data = await ideal_image.read()

        if len(current_data) == 0 or len(ideal_data) == 0:
            return JSONResponse(
                status_code=400,
                content=ErrorResponse(
                    error=ErrorDetail(
                        code=ErrorCodes.VALIDATION_ERROR,
                        message="Image files are required",
                    )
                ).model_dump(),
            )

        _, cv2_current = validate_image(current_data)
        _, cv2_ideal = validate_image(ideal_data)

        # Perform staged morphing
        morph_service = get_morphing_service()
        results = morph_service.morph_stages(
            cv2_current,
            cv2_ideal,
            stage_values,
            img1_label="current",
            img2_label="ideal",
        )

        # Convert results
        stage_images = []
        dimensions = None

        for progress, img in results:
            h, w = img.shape[:2]
            if dimensions is None:
                dimensions = ImageDimensions(width=w, height=h)

            image_base64 = cv2_to_base64(img, "png")
            stage_images.append(StageImage(progress=progress, image=image_base64))

        # Calculate processing time
        processing_time_ms = (time.time() - start_time) * 1000

        return StagedMorphResponse(
            data=StagedMorphData(
                images=stage_images,
                format="png",
                dimensions=dimensions or ImageDimensions(width=0, height=0),
            ),
            meta=ResponseMeta(
                request_id=request_id,
                processing_time_ms=processing_time_ms,
            ),
        )

    except ImageValidationError as e:
        status_code_map = {
            ErrorCodes.IMAGE_TOO_LARGE: 413,
            ErrorCodes.INVALID_IMAGE_FORMAT: 400,
            ErrorCodes.FACE_NOT_DETECTED: 400,
            ErrorCodes.MULTIPLE_FACES: 400,
        }
        status_code = status_code_map.get(e.code, 400)

        return JSONResponse(
            status_code=status_code,
            content=ErrorResponse(
                error=ErrorDetail(
                    code=e.code,
                    message=e.message,
                    details=e.details,
                )
            ).model_dump(),
        )

    except Exception as e:
        print(f"Morph stages error: {e}")

        return JSONResponse(
            status_code=500,
            content=ErrorResponse(
                error=ErrorDetail(
                    code=ErrorCodes.PROCESSING_ERROR,
                    message="Failed to generate morphed images",
                )
            ).model_dump(),
        )

File: frontend/src/lib/api/morph.ts
/**
 * モーフィングAPI
 */

import { apiPostFormData } from './client';
import { MorphData, StagedMorphData, PartsBlendData, PartsSelection } from './types';

/**
 * デフォルトの段階値
 */
export const DEFAULT_STAGES = [0, 0.25, 0.5, 0.75, 1.0];

/**
 * 2つの顔画像をモーフィングする
 *
 * @param current - 現在の顔画像（JPEG/PNG、最大10MB）
 * @param ideal - 理想の顔画像（JPEG/PNG、最大10MB）
 * @param progress - 変化度合い（0.0 = 現在、1.0 = 理想、デフォルト: 0.5）
 * @returns モーフィング結果（Base64エンコードされた画像）
 * @throws ApiError - 顔未検出、画像フォーマットエラーなど
 *
 * @example
 * ```typescript
 * try {
 *   const result = await morphImages(currentFile, idealFile, 0.5);
 *   const imgSrc = `data:image/png;base64,${result.image}`;
 *   document.querySelector('img')!.src = imgSrc;
 * } catch (error) {
 *   console.error('Morphing failed:', error);
 * }
 * ```
 */
export async function morphImages(
  current: File,
  ideal: File,
  progress: number = 0.5
): Promise<MorphData> {
  const formData = new FormData();
  formData.append('current_image', current);
  formData.append('ideal_image', ideal);
  formData.append('progress', progress.toString());

  return apiPostFormData<MorphData>('/api/v1/morph', formData);
}

/**
 * 2つの顔画像を複数段階でモーフィングする
 *
 * @param current - 現在の顔画像（JPEG/PNG、最大10MB）
 * @param ideal - 理想の顔画像（JPEG/PNG、最大10MB）
 * @param stages - 生成する段階の配列（デフォルト: [0, 0.25, 0.5, 0.75, 1.0]）
 * @returns 段階別モーフィング結果
 * @throws ApiError - 顔未検出、画像フォーマットエラーなど
 *
 * @example
 * ```typescript
 * try {
 *   const result = await morphStages(currentFile, idealFile);
 *   result.images.forEach(({ progress, image }) => {
 *     console.log(`Progress ${progress * 100}%:`, image.substring(0, 50));
 *   });
 * } catch (error) {
 *   console.error('Staged morphing failed:', error);
 * }
 * ```
 */
export async function morphStages(
  current: File,
  ideal: File,
  stages: number[] = DEFAULT_STAGES
): Promise<StagedMorphData> {
  const formData = new FormData();
  formData.append('current_image', current);
  formData.append('ideal_image', ideal);
  formData.append('stages', JSON.stringify(stages));

  // 段階的モーフィングは時間がかかるため、長めのタイムアウトを設定
  return apiPostFormData<StagedMorphData>('/api/v1/morph/stages', formData, {
    timeout: 60000, // 60秒
  });
}

/**
 * Base64画像データをDataURLに変換するユーティリティ
 *
 * @param base64 - Base64エンコードされた画像データ
 * @param format - 画像フォーマット（デフォルト: 'png'）
 * @returns DataURL形式の文字列
 *
 * @example
 * ```typescript
 * const result = await morphImages(currentFile, idealFile);
 * const dataUrl = toDataUrl(result.image);
 * document.querySelector('img')!.src = dataUrl;
 * ```
 */
export function toDataUrl(base64: string, format: string = 'png'): string {
  // 既にdata:形式の場合はそのまま返す
  if (base64.startsWith('data:')) {
    return base64;
  }
  return `data:image/${format};base64,${base64}`;
}

/**
 * ブレンドメソッドの型
 */
export type BlendMethod = '2d' | '3d' | 'auto';

/**
 * 理想の顔から選択したパーツを現在の顔にブレンドする
 *
 * @param current - 現在の顔画像（JPEG/PNG、最大10MB）
 * @param ideal - 理想の顔画像（JPEG/PNG、最大10MB）
 * @param parts - ブレンドするパーツの選択
 * @param method - ブレンド方式（'2d', '3d', 'auto'）。デフォルトは'auto'
 *                 'auto'は3Dが利用可能な場合は3D、それ以外は2Dを使用
 * @returns ブレンド結果（Base64エンコードされた画像）
 * @throws ApiError - 顔未検出、画像フォーマットエラーなど
 *
 * @example
 * ```typescript
 * try {
 *   const result = await blendParts(currentFile, idealFile, {
 *     left_eye: true,
 *     right_eye: true,
 *     nose: true,
 *     lips: false,
 *     left_eyebrow: false,
 *     right_eyebrow: false,
 *   }, '3d');
 *   const imgSrc = `data:image/png;base64,${result.image}`;
 *   document.querySelector('img')!.src = imgSrc;
 * } catch (error) {
 *   console.error('Parts blend failed:', error);
 * }
 * ```
 */
export async function blendParts(
  current: File,
  ideal: File,
  parts: PartsSelection,
  method: BlendMethod = 'auto'
): Promise<PartsBlendData> {
  const formData = new FormData();
  formData.append('current_image', current);
  formData.append('ideal_image', ideal);
  formData.append('parts', JSON.stringify(parts));
  formData.append('method', method);

  // 3Dブレンドは特に時間がかかるため、より長めのタイムアウトを設定
  return apiPostFormData<PartsBlendData>('/api/v1/blend/parts', formData, {
    timeout: 120000, // 120秒（3Dは処理に時間がかかる）
  });
}
